python/GafferUI/NumericWidget.py

```diff
@@ -304,3 +312,70 @@ def __emitValueChanged( self, reason ) :
 			return
 
 		signal( self, reason )
+
+# A basic validator/evaluator that supports simple maths
+# operators [+-/*] along with standard number validation, eg:
+#   2 + 3
+#   4.4 / 2
+class BasicNumericMathExpression( QtGui.QValidator ) :
+
+	__invalid = re.compile( r"[^0-9\.+\-*/\s]" )
+	__operators = re.compile( r"[+\-/*]" )
+	__whitespace = re.compile( r"\s+" )
+
+	def __init__( self, parent, typeValidator ) :
+
+		QtGui.QValidator.__init__( self, parent )
+
+		self.__typeValidator = typeValidator
+		self.__typeValidator.setParent( self )
+
+	def validate( self, text, pos ) :
+
+		if not self.__isExpression( text ) :
+			return self.__typeValidator.validate( text, pos )
+
+		# Invalid chars
+		if re.search( self.__invalid, text ) is not None :
+			return QtGui.QValidator.Invalid, text, pos
+
+		# Remove whitespace to simplify tokenisation
+		text = re.sub( self.__whitespace, "", text )
+
+		# Ends with an operator
+		if re.match( self.__operators, text[ -1 ] ) :
+			return QtGui.QValidator.Intermediate, text, pos
+
+		# Individual numbers invalid for our type validator
+		for number in re.split( self.__operators, text ) :
+			# May be seen with *-1, etc...
+			if number == '' :
+				continue
+			state = self.__typeValidator.validate( number, len(number)-1 )
+			if state[0] != QtGui.QValidator.Acceptable :
+				return state[0], text, pos
+
+		# Some other syntax error
+		try :
+			self.eval( text )
+		except SyntaxError :
+			return QtGui.QValidator.Invalid, text, pos
```

Comment:
I can't but feel that this could be simpler. We're mixing explicit `__isExpression()` checks, string substitutions, manual splitting , regexes, preexisting validators _and_ calls to `eval()`. Can this just be based around a single `(.+)([+-/*])?(.+)?` regex?

Reply:
How about [this](https://github.com/GafferHQ/gaffer/pull/3873/commits/4505a343edc81a5073011877110c4b503c8b1efa). It also adds `fixup`, which prevents people leaving the field with a trailing operator. There is however a [bug](https://stackoverflow.com/questions/34055174/qvalidator-fixup-issue#34056852), in the `1.6.0` deps version of the Qt bindings that prevents `fixup` from doing anything (we don't get the `QString` there, so its a lost cause). Fortunately, this is resolved in the `2.0.0` deps, so I've re-based this to `master`.

Reply:
I’m not following the implementation details too closely, but it’d be a pity to defer this to master rather than 0.58, purely to avoid a bug that isn’t todo with Gaffer itself. At IE we’d expect to be able to deploy 0.59 using Qt5.6 anyhow, so for us, the bug won’t even be fixed in 0.59.

Reply:
In short, you can't stop people leaving trailing operators/invalid expressions in the field without a newer version of `PySide`, so its up to you you and @johnhaddon if you want to have that behaviour in Gaffer or not...

Reply:
I guess a more positive framing is that I’d hope to be able to deploy Gaffer 0.58 with a more modern Qt/PySide and benefit from both the feature and the fix.

That said, it’s obviously not production critical to get this soon, just a nice thing.

Reply:
I think my main concern is that we'd be leaving in a known issue in stock gaffer, which is used across the entirety of CS and other sites just to support custom builds.

Reply:
From my perspective, we leave in hundreds of known issues in every release of Gaffer. At least this one would already have fix in a future release or a custom build of the same release.

Reply:
I've added d7c81c8 which simplifies the logic and adds a workaround for the PySide bug. We'll release this in 0.58.